File: Production/lib/library_panel_contract.py
"""LIBRARY_PANEL_CLASSIFICATION_V1 — server authority for library panel tabs.

Every GET /api/cr/library row carries ``panel_tabs`` so the client never
infers visibility from Directus ``asset_type`` or duplicated tier maps.
"""

from __future__ import annotations

from typing import Any

# Mirrors LibraryPanel.tsx LIBRARY_TIERS (Kim BS3 lock 2026-05-06).
LIBRARY_PANEL_TABS: tuple[str, ...] = (
    "images",
    "ambient",
    "sfx",
    "transitions",
    "watercolors",
)

_IMAGES_DISK_TIERS = frozenset({"source", "cropped", "character_master", "element_pose"})


def panel_tabs_for_cr_library_row(item: dict[str, Any]) -> list[str]:
    """Return panel tab ids for one cr_library list row (may be empty)."""
    tier = str(item.get("tier") or "")
    if tier == "canonical":
        return []
    if tier == "watercolor":
        return ["watercolors"]
    if tier in _IMAGES_DISK_TIERS:
        return ["images"]
    tags = item.get("tags") or []
    if isinstance(tags, list) and "watercolor" in tags:
        return ["watercolors"]
    asset_type = str(item.get("asset_type") or "")
    if asset_type in ("watercolor_static", "watercolor"):
        return ["watercolors"]
    if asset_type in ("audio",):
        return ["ambient"] if "ambient" in tags else []
    if asset_type in ("sfx", "transition"):
        return ["sfx"] if asset_type == "sfx" else ["transitions"]
    # Unknown disk tier — default to images so rows are never silently hidden.
    if tier:
        return ["images"]
    return []


def attach_panel_tabs(item: dict[str, Any]) -> dict[str, Any]:
    """Mutate and return *item* with ``panel_tabs`` set from disk-scan tier."""
    tabs = panel_tabs_for_cr_library_row(item)
    if tabs:
        item["panel_tabs"] = tabs
    else:
        item.pop("panel_tabs", None)
    return item


def attach_panel_tabs_all(items: list[dict[str, Any]]) -> None:
    for item in items:
        attach_panel_tabs(item)


def row_matches_panel_filter(item: dict[str, Any], panel: str) -> bool:
    tabs = item.get("panel_tabs")
    if isinstance(tabs, list) and tabs:
        return panel in tabs
    return panel in panel_tabs_for_cr_library_row(item)

File: Production/tools/storyboard-v2/src/utils/libraryPanelContract.ts
/** LIBRARY_PANEL_CLASSIFICATION_V1 — client filter uses server panel_tabs authority. */

export type LibraryPanelTab =
  | 'images'
  | 'ambient'
  | 'sfx'
  | 'transitions'
  | 'watercolors';

export interface LibraryPanelRow {
  panel_tabs?: string[];
  tier?: string;
  asset_type?: string;
  tags?: string[];
}

/** Primary filter: server-emitted panel_tabs. Legacy tier/asset_type fallback only. */
export function libraryItemMatchesPanelTab(
  it: LibraryPanelRow,
  tab: LibraryPanelTab,
): boolean {
  const tabs = it.panel_tabs;
  if (Array.isArray(tabs) && tabs.length > 0) {
    return tabs.includes(tab);
  }
  return legacyLibraryItemMatchesPanelTab(it, tab);
}

function legacyLibraryItemMatchesPanelTab(
  it: LibraryPanelRow,
  tab: LibraryPanelTab,
): boolean {
  const t = it.tier ?? '';
  if (tab === 'images') {
    if (t === 'canonical') return false;
    if (
      t === 'source'
      || t === 'cropped'
      || t === 'character_master'
      || t === 'element_pose'
    ) {
      return true;
    }
    const at = inferLegacyAssetType(it);
    if (at === 'canonical_image') return false;
    return (
      at === 'image'
      || at === 'still_delivery'
      || at === 'still_master'
      || at === 'beat_scene'
      || at === 'element_pose'
    );
  }
  if (tab === 'watercolors') {
    if (t === 'watercolor') return true;
    const tags = it.tags ?? [];
    if (tags.includes('watercolor')) return true;
    return inferLegacyAssetType(it) === 'watercolor_static';
  }
  if (tab === 'ambient') {
    const tags = it.tags ?? [];
    return inferLegacyAssetType(it) === 'audio' && tags.includes('ambient');
  }
  if (tab === 'sfx') {
    const at = inferLegacyAssetType(it);
    return at === 'sfx' || at === 'transition';
  }
  if (tab === 'transitions') {
    return (it.tags ?? []).includes('transition');
  }
  return false;
}

function inferLegacyAssetType(it: LibraryPanelRow): string {
  if (it.asset_type) return it.asset_type;
  if (it.tier === 'canonical') return 'canonical_image';
  if (it.tier === 'element_pose') return 'element_pose';
  if (it.tier === 'character_master') return 'still_master';
  if (it.tier === 'source' || it.tier === 'cropped') return 'still_delivery';
  return 'image';
}

/** Stitch editor audio rows — panel_tabs assigned client-side until API emits them. */
export function stitchAudioPanelTabs(category: string): string[] {
  if (category === 'ambient') return ['ambient'];
  if (category === 'sfx') return ['sfx'];
  if (category === 'transitions') return ['transitions'];
  return [];
}
